### ISSUE TITLE
Using client tokens to access dynamodb from lambda?

### ISSUE THREAD

**Issue by ToddHoff** [aws-amplify/amplify-js#2682]
** Which Category is your question related to? **
Credentials, DynamoDB

** What AWS Services are you utilizing? **
Cognito, Lambda, DynamoDB

** Provide additional details e.g. code snippets **

My end goal is to use fine-grained permission so I can restrict record access to the user that created it. 

My first step is in the web client I have the the id, access, and refresh tokens. I want to send the tokens from the client to the lambda function and reestablish the session such that when I made the dynamodb call this iam policy will work:

"Condition": {
                "ForAllValues:StringEquals": {
                    "dynamodb:LeadingKeys": ["${cognito-identity.amazonaws.com:sub}"]
                }
            }
        }

What I'm unsure of is how to make this happen? This is what I have so far, but it seems 
complex and redundant. I have the tokens, what else should I need? But if i don't do the 
refresh AWS.config.credentials.identityId doesn't get set, which I assume must be necessary
for the i am to work.

 async authenticateWithTokens(tokens) {
     // the tokens are passed in from the web client via the lambda invoke

      var self = this;

      const AccessToken = new AmazonCognitoIdentity.CognitoAccessToken({ AccessToken: tokens.accessToken });
      const IdToken = new AmazonCognitoIdentity.CognitoIdToken({ IdToken: tokens.idToken });
      const RefreshToken = new AmazonCognitoIdentity.CognitoRefreshToken({ RefreshToken: tokens.refreshToken });

      const session_data = {
        IdToken: IdToken,
        AccessToken: AccessToken,
        RefreshToken: RefreshToken
      };
    
      const user_session = new AmazonCognitoIdentity.CognitoUserSession(session_data);

      const user_data = {
         Username: "email@gmail.com",
         Pool: this.userPool
      };

      const cognito_user = new AmazonCognitoIdentity.CognitoUser(user_data);
      cognito_user.setSignInUserSession(user_session);

      const promise = new Promise((resolve, reject) => {

         cognito_user.getSession(function (error, session) { 

            if (session.isValid()) {

               self.cfg.AWS.config.credentials = self.getCognitoIdentityCredentials(tokens);

               self.cfg.AWS.config.credentials.refresh((error) => {

                  if (error) {
                     reject(error);
                  } else {
                     log.info("!!!ID:" + self.cfg.AWS.config.credentials.identityId);
                     resolve(null);
                 }
              })
            }
    else {
               reject(error);
            }
          })
      })

      return promise;

   }// authenticateWithTokens

Any help would be greatly appreciated.

**Comment by ToddHoff** [comment]
To skip the token part and passed the user name and password from the client to lambda function, authenticated in the lambda function, and it didn't work. BTW, is there a problem with this approach in general? This doesn't seem less secure than passing tokens.

This is what I added to my dnynamodb IAM role for the User table. The User table id is the federated id.

They key is:
us-east-1:cbd72a19-787a-4314-a5d7-0c695bc89657

It's pulled from AWS.config.credentials.identityId after the authenticated cognito with the username and password.

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "dynamodb:DeleteItem",
                "dynamodb:GetItem",
                "dynamodb:PutItem",
                "dynamodb:Query",
                "dynamodb:UpdateItem"
            ],
            "Resource": [
                "arn:aws:dynamodb:*:*:table/User"
            ],
            "Condition": {
                "ForAllValues:StringEquals": {
                    "dynamodb:LeadingKeys": [
                        "${cognito-identity.amazonaws.com:sub}"
                    ]
                }
            }
        }
    ]
}

I got this error:
There was a problem: User: arn:aws:sts::369659351212:assumed-role/redBasisCalcRole/redBasisCalc is not authorized to perform: dynamodb:GetItem on resource: arn:aws:dynamodb:us-east-1:369659351212:table/User

Then at other items the get with the document client succeeded, but there were no attributes, which means the record wasn't found. 

So I'm confused and at a loss at how to debug this. Are there any logs that can be generated so I can see where it is going wrong?

**Comment by ToddHoff** [comment]
So trying once more, but this time with some progress...

User: arn:aws:sts::369659351212:assumed-role/Cognito_BasisAuth_Role/CognitoIdentityCredentials is not authorized to perform: dynamodb:PutItem on resource: arn:aws:dynamodb:us-east-1:369659351212:table/User

I noticed that it mentioned Cognito_BasisAuth_Role. I had put the clause

Condition": {
"ForAllValues:StringEquals": {
"dynamodb:LeadingKeys": ["${cognito-identity.amazonaws.com:sub}"]
}

on the lambda function itself, not the cognito's authenticated role.  When I added it to that role and i created the client like:

new AWS.DynamoDB.DocumentClient({ credentials: AWS.config.credentials });

 it worked.

I'm confused. I thought the effective permissions was governed by the lambda function's IAM role. What am I missing?
